table/format.cc

```diff
@@ -176,64 +187,80 @@ void Footer::set_table_magic_number(uint64_t magic_number) {
+// Footer format, in three parts:
+// * Part1
+//   -> format_version == 0 (inferred from legacy magic number)
+//      <empty> (0 bytes)
+//   -> format_version >= 1
+//      checksum type (char, 1 byte)
+// * Part2
+//      metaindex handle (varint64 offset, varint64 size)
+//      index handle     (varint64 offset, varint64 size)
+//      <zero padding> for part2 size = 2 * BlockHandle::kMaxEncodedLength = 40
+// * Part3
+//   -> format_version == 0 (inferred from legacy magic number)
+//      legacy magic number (8 bytes)
+//   -> format_version >= 1 (inferred from NOT legacy magic number)
+//      format_version (uint32LE, 4 bytes), also called "footer version"
+//      newer magic number (8 bytes)
+void Footer::EncodeTo(std::string* dst, uint64_t footer_offset) const {
+  (void)footer_offset;  // Future use
+
+  // Sanitize magic numbers & format versions
+  assert(table_magic_number_ != kNullTableMagicNumber);
+  uint64_t magic = table_magic_number_;
+  uint32_t fv = format_version_;
+  assert(fv != kInvalidFormatVersion);
+  assert(IsLegacyFooterFormat(magic) == (fv == 0));
+
+  ChecksumType ct = checksum_type();
+
+  // Allocate destination data and generate parts 1 and 3
+  const size_t original_size = dst->size();
+  char* part2;
+  if (fv > 0) {
+    dst->resize(original_size + kNewVersionsEncodedLength);
+    char* part1 = &(*dst)[original_size];
+    part2 = part1 + 1;
+    char* part3 = part2 + 2 * BlockHandle::kMaxEncodedLength;
+    assert(&(*dst)[dst->size() - 1] + 1 - part3 == /* part 3 size */ 12);
+    // Generate parts 1 and 3
+    part1[0] = ct;
+    EncodeFixed32(part3, fv);
+    EncodeFixed64(part3 + 4, magic);
   } else {
-    const size_t original_size = dst->size();
-    dst->push_back(static_cast<char>(checksum_));
-    metaindex_handle_.EncodeTo(dst);
-    index_handle_.EncodeTo(dst);
-    dst->resize(original_size + kNewVersionsEncodedLength - 12);  // Padding
-    PutFixed32(dst, version());
-    PutFixed32(dst, static_cast<uint32_t>(table_magic_number() & 0xffffffffu));
-    PutFixed32(dst, static_cast<uint32_t>(table_magic_number() >> 32));
-    assert(dst->size() == original_size + kNewVersionsEncodedLength);
+    dst->resize(original_size + kVersion0EncodedLength);
+    part2 = &(*dst)[original_size];
+    char* part3 = part2 + 2 * BlockHandle::kMaxEncodedLength;
+    assert(&(*dst)[dst->size() - 1] + 1 - part3 == /* part 3 size */ 8);
+    // Legacy SST files use kCRC32c checksum but it's not stored in footer.
+    assert(ct == kNoChecksum || ct == kCRC32c);
+    // Generate part 3 (part 1 empty)
+    EncodeFixed64(part3, magic);
   }
-}
 
-Footer::Footer(uint64_t _table_magic_number, uint32_t _version)
-    : version_(_version),
-      checksum_(kCRC32c),
-      table_magic_number_(_table_magic_number) {
-  // This should be guaranteed by constructor callers
-  assert(!IsLegacyFooterFormat(_table_magic_number) || version_ == 0);
+  // Generate Part2
+  // Variable size encode handles (sigh)
```

Comment:
I never noticed before we use varint and then zero padding to make it a fixed size.